cedar-lean/Cedar/Thm/Lemmas/Typechecker.lean

```diff
@@ -0,0 +1,167 @@
+        rty.find? k = some qty → r.find? k = some v → InstanceOfType v qty.getType) →
+      -- required attributes are present
+      (∀ (k : Attr) (qty : QualifiedType), rty.find? k = some qty → qty.isRequired → r.contains k) →
+      InstanceOfType (.record r) (.record rty)
+  | instance_of_ext (e : Ext) (ety : ExtType) : InstanceOfExtType e ety →
+      InstanceOfType (.ext e) (.ext ety)
+
+def InstanceOfRequestType (request : Request) (reqty : RequestType) : Prop :=
+  InstanceOfEntityType request.principal reqty.principal ∧
+  request.action = reqty.action ∧
+  InstanceOfEntityType request.resource reqty.resource ∧
+  InstanceOfType request.context (.record reqty.context)
+
+/--
+For every entity in the store,
+1. The entity's type is defined in the type store.
+2. The entity's attributes match the attribute types indicated in the type store.
+3. The entity's ancestors' types are defined in the type store, and the ancestor
+   relationship is consistent with the descendants information in the type store.
+-/
+def InstanceOfEntityTypeStore (entities : Entities) (ets: EntityTypeStore) : Prop :=
+  ∀ uid data, entities.find? uid = some data →
+    ∃ attrTys descendantTys, ets.find? uid.ty = some (attrTys, descendantTys) ∧
+      InstanceOfType data.attrs (.record attrTys) ∧
+        ∀ ancestor, ancestor ∈ data.ancestors →
+          ∃ descendants, ets.find? ancestor.ty = some descendants ∧ uid.ty ∈ descendantTys
+
+/--
+For every action in the entity store, the action's ancestors are defined in the
+action store, and the ancestor relationships in the entity store are consistent
+with the descendants information in the action store.
+-/
+def InstanceOfActionStore (entities : Entities) (as: ActionStore) : Prop :=
+  ∀ uid data, entities.find? uid = some data → uid.isAction →
+    ∀ ancestor, ancestor ∈ data.ancestors →
+      ∃ descendants, as.find? ancestor = some descendants ∧ uid ∈ descendants
+
+def RequestAndEntitiesMatchEnvironment (env : Environment) (request : Request) (entities : Entities) : Prop :=
+  InstanceOfRequestType request env.reqty ∧
+  InstanceOfEntityTypeStore entities env.ets ∧
+  InstanceOfActionStore entities env.acts
+
+/--
+The type soundness property says that if the typechecker assigns a type to an
+expression, then it must be the case that the expression `EvaluatesTo` a value
+of that type. The `EvaluatesTo` predicate covers the (obvious) case where evaluation
+has no errors, but it also allows for an error of type `entityDoesNotExist`.
+
+The typechecker cannot protect against `entityDoesNotExist` errors because they
+depend on runtime information (i.e., the entities passed into the authorization
+request). All other errors (`attrDoesNotExist`, `typeError`, and `extensionError`)
+can be prevented statically. `extensionError` errors can be prevented because
+validation only allows extension function constructors to apply to string literals,
+meaning that they can be fully evaluated during validation. Currently, only
```

Comment:
Should we allow for extension errors too, in case a future (non-constructor) extension function can error?  

(I'm not sure how much code we'd have to change to allow for this in the future.)

Reply:
Sure, I can make that change. If wanted, we can add a separate proof later that `extensionError`s are (currently) disallowed.